docs/architecture/psa-migration/md-cipher-dispatch.md

```diff
@@ -0,0 +1,450 @@
+
+#### Conversion to PSA encoding
+
+The implementation needs to convert from a legacy type encoding to a PSA encoding.
+
+```
+static inline psa_algorithm_t psa_alg_of_md_info(
+    const mbedtls_md_info_t *md_info );
+```
+
+#### Determination of PSA support at runtime
+
+```
+int psa_can_do_hash(psa_algorithm_t hash_alg);
+```
+
+The job of this private function is to return 1 if `hash_alg` can be performed through PSA now, and 0 otherwise. It is only defined on algorithms that are enabled via PSA.
+
+As a starting point, return 1 if PSA crypto has been initialized. This will be refined later (to return 1 if the [accelerator subsystem](https://github.com/Mbed-TLS/mbedtls/issues/6007) has been initialized).
+
+Usage note: for algorithms that are not enabled via PSA, calling `psa_can_do_hash` is generally safe: whether it returns 0 or 1, you can call a PSA hash function on the algorithm and it will return `PSA_ERROR_NOT_SUPPORTED`.
+
+#### Support for PSA dispatch in hash operations
+
+Each function that performs some hash operation or context management needs to know whether to dispatch via PSA or legacy.
+
+If given an established context, use its `engine` field.
+
+If given an algorithm as an `mbedtls_md_type_t type` (possibly being the `type` field of a `const mbedtls_md_info_t *`):
+
+* If there is a PSA accelerator for this hash and `psa_can_do_hash(alg)`, call the corresponding PSA function, and if applicable set the engine to `MBEDTLS_MD_ENGINE_PSA`. (Skip this is `MBEDTLS_MD_SOME_PSA` is not defined.)
+* Otherwise dispatch to the legacy module based on the type as currently done. (Skip this is `MBEDTLS_MD_SOME_LEGACY` is not defined.)
+* If no dispatch is possible, return `MBEDTLS_ERR_MD_FEATURE_UNAVAILABLE`.
+
+Note that this assumes that an operation that has been started via PSA can be completed. This implies that `mbedtls_psa_crypto_free` must not be called while an operation using PSA is in progress. Document this.
+
+#### Error code conversion
+
+After calling a PSA function, call `mbedtls_md_error_from_psa` to convert its status code. This function is currently defined in `hash_info.c`.
+
+### Migration to MD light
+
+#### Migration of modules that used to call MD and now do the legacy-or-PSA dance
+
+Get rid of the case where `MBEDTLS_MD_C` is undefined. Enable `MBEDTLS_MD_LIGHT` in `build_info.h`.
+
+#### Migration of modules that used to call a low-level hash module and now do the legacy-or-PSA dance
+
+Switch to calling MD (light) unconditionally. Enable `MBEDTLS_MD_LIGHT` in `build_info.h`.
+
+#### Migration of modules that call a low-level hash module
+
+Switch to calling MD (light). Enable `MBEDTLS_MD_LIGHT` in `build_info.h`.
+
+#### Migration of use-PSA mixed code
+
+Instead of calling `hash_info.h` functions to obtain metadata, get it from `md.h`.
+
+Optionally, code that currently tests on `MBEDTLS_USE_PSA_CRYPTO` just to determine whether to call MD or PSA to calculate hashes can switch to just having the MD variant.
```

Comment:
I think that's something that we may have to decide rather than leave as optional. More precisely, I think the following two questions are related:

1. For code that currently calls MD or PSA depending on whether `MBEDTLS_USE_PSA_CRYPTO` is defined, do we want to (a) keep it as it is, or (b) switch to just having the MD variant?
2. Do we want to (a) support configurations where a hash is only available in legacy but not PSA, or (b) forbid that (most probably by auto-enabling things)?

If we answer 1a and 2a, then in a configuration where (for example) `MBEDTLS_SHA384_C` is defined but `PSA_WANT_SHA_384` is not, then TLS can't use SHA-384. Since `MBEDTLS_MD_CAN_SHA384` would be defined in this build, that means TLS (or code governed by `USE_PSA`) would need its own set of test macros for hashes.

So, we'd find ourselves again in a situation where you have multiple sets of hash-support macros and you need to remember to use the right one for the right module, which is one of the issues with the current design that we're trying to avoid. Granted, it would much less bad because we'd get fewer family of macros, and perhaps more importantly deciding which family to pick wouldn't require looking at the source code, basically there would be one family for code governed by `USE_PSA` and one for the rest, but which part of the code is governed by `USE_PSA` is a bit trickier to explain that we'd like to (for example, TLS 1.3 is not governed by `USE_PSA` except for the parts that are shared with TLS 1.2).

Note: answering 1b would only partially solve the matter because we'd still have to be careful about TLS: which parts are 1.3-only and which parts are 1.2-or-common. So, I think I'd really like to answer 2b so that we avoid that kind of problem altogether.

Actually, thinking while writing: if we pick 2b, then isn't `MBEDTLS_MD_CAN_xxx` synonymous with `PSA_WANT_ALG_xxx`?

Reply:
> For code that currently calls MD or PSA depending on whether MBEDTLS_USE_PSA_CRYPTO is defined, do we want to (a) keep it as it is, or (b) switch to just having the MD variant?

This doesn't matter much since MD can always do everything PSA can do. It only makes a difference if it's possible to have algorithms available through the legacy interface but not PSA (beyond the case where PSA is disabled), in  which case going through MD when `MBEDTLS_USE_PSA_CRYPTO` is defined could make some parts of the library support extra algorithms. I guess there's a risk that this support will be inconsistent, which is an argument to either keep the current compilation guards or add the requirement that all legacy algorithms are available through PSA.

> Do we want to (a) support configurations where a hash is only available in legacy but not PSA, or (b) forbid that (most probably by auto-enabling things)?

I haven't made it a requirement that legacy support implies PSA support (assuming `MBEDTLS_PSA_CRYPTO_C` is enabled) because it wasn't actually necessary. But even if it doesn't seem to be necessary, I'm ok with it as a simplifying assumption, which reduces the risk that we miss something.

> if we pick 2b, then isn't MBEDTLS_MD_CAN_xxx synonymous with PSA_WANT_ALG_xxx?

Only if `defined(MBEDTLS_PSA_CRYPTO_C)`.

Reply:
> As a developer of library code that uses Mbed TLS to perform cryptographic operations,  
I want to know which functions to call and which feature macros to check,  
so that my code works in all Mbed TLS configurations.

As such a developer (or, almost equivalently, as a contributor to Mbed TLS writing tests), I want to know which feature macro to check to see if SHA-384 is supported in TLS 1.2. If things stay as they are (ie, 1a and 2a above), then I have to do:
```
#if (!defined(MBEDTLS_USE_PSA_CRYPTO) && defined(MBEDTLS_SHA384_C)) ||
     (defined(MBEDTLS_USE_PSA_CRYPTO) && defined(PSA_WANT_ALG_SHA_384)
```
unless there's a macro `MBEDTLS_TLS12_CAN_SHA384` or `MBEDTLS_CODE_THAT_OBEYS_USE_PSA_CRYPTO_CAN_SHA384`.

Regardless, I have to pick the right condition - what about TLS 1.3? Well, for the most part it's `PSA_WANT_ALG_SHA_384` but for the running handshake hash it's `MBEDTLS_SHA384_C` when `MBEDTLS_USE_PSA_CRYPTO` is not defined, so I need to check both...

My point is twofold:

- If we don't add a requirement that legacy support implies PSA support, then whether the code in TLS (and others) keeps doing the current `#if defined(MBEDTLS_USE_PSA_CRYPTO)` dance or switches to just using MD is user-visible (in those rares configs that we keep allowing), so I don't think that's something we can leave as "Optional" as the document states, we need to make a decision.
- The decision should make things as simple as possible for users (and ourselves).

Unless I missed something, we can achieve a state where users can just use `MBEDTLS_MD_CAN_xxx` as a universal feature-check macro for hashing by doing one (or both) of the following:
- (1b above) change all the code that currently does the `#if defined(MBEDTLS_USE_PSA_CRYPTO)` dance to just always use MD instead; or
- (2b above) make it a requirement that legacy support implies PSA support (assuming `MBEDTLS_PSA_CRYPTO_C` is enabled), because then the set of hashes support by MD and the set of hashes supported by PSA are equal.

I think (1b) has drawbacks:
- it's actually not just the code that does the `#if defined(MBEDTLS_USE_PSA_CRYPTO)` dance, we'd need to also switch TLS 1.3 code to use MD in order for things to be really uniform;
- it would mean removing the PSA-calling path from the code base, only to have to write it again later.

Obviously, (2b) has drawbacks as well, in that it may increase code size in some cases as we discussed (like, you start enabling hash XXX in PSA, but you have an RSA-PSS accelerator that doesn't support it, so now the built-in RSA gets included even though you might not care about RSA-PSS-XXX), but the configs that would be rule out are things we supported rather by accident and were not planning to keep supporting in a PSA-only future anyway.

So overall, I'm more inclined towards (2b), mostly because it makes reasoning about things easier: less unusual configs to consider (and ideally test, which I don't think we are doing so far). Wdyt?